File: 0881-boats-to-save-people/0881-boats-to-save-people.py
from collections import deque
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        if max(people) > limit:
            return 0
        people.sort()
        people = deque(people)
        ans = 0
        while people:
            temp = 0
            cnt = 0
            while people and temp + people[-1] <= limit and cnt < 2:
                temp += people.pop()
                cnt += 1
    
            while people and temp + people[0] <= limit and cnt < 2:
                temp += people.popleft()
                cnt += 1
            ans += 1
        return ans

File: 0881-boats-to-save-people/0881-boats-to-save-people.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn num_rescue_boats(people: Vec<i32>, limit: i32) -> i32 {
        let mut people = people.clone();
        
        people.sort_by(|a, b| b.cmp(&a));
    
        let mut h: BinaryHeap<(i32, i32)> = BinaryHeap::new();
        let mut over = 0;
        
        for weight in people {
            if let Some((min, cnt)) = h.pop() {                
                let min = -min;
                
                if min + weight <= limit {
                    over += 1;
                } else {
                    h.push((-min, cnt));
                    h.push((-weight, 1));
                }
            } else {
                h.push((-weight, 1));
            }
        }
        over + h.len() as i32
    }
}